package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -548,14 +533,21 @@ cdef class FastNS(object):
             raise ValueError("search_coords must have a shape of (n, 3), got "
                              "{}.".format(search_coords.shape))
 
+        tmpcoords = search_coords.copy()
```

Comment:
I'm not a huge fan of doing a numpy copy here. Can we keep the wrapping of individual positions inside the loop (and call with n=1)?

Reply:
Yeah I thought about this, too. I did this because I felt a bit uncomfortable with having a function call in a loop, and I thought that an O(n)  memory footprint should be affordable. However, if the function call gets inlined by the compiler, copying the array will certainly be slower. So, in the end, you are probably right but we'll have to measure the difference to be sure.

Reply:
Ah, there's one more thing: Distinguishing between orthorhombic and general triclinic coordinate folding potentially makes the code a bit faster, and I wanted to avoid checking for `self.triclinic` in the loop. For large systems, branch prediction will most likely eliminate the impact, but it might make function inlining less likely (that's purely speculative, though).

Reply:
BTW I think this is not a numpy copy operation but a to-C-contiguous [Memoryview copy](https://cython.readthedocs.io/en/latest/src/userguide/memoryviews.html#c-and-fortran-contiguous-copies). Don't know if there's a huge difference since both operations require the GIL. Using `tmpcoords[:, :] = search_coords` does not require the GIL but will throw a `ValueError` at runtime if `search_coords` is non-contiguous, so that's not an option.

Reply:
It's more I'd rather have 3 floats on the stack than O(n) on the heap

Reply:
Is it because you want to avoid the entailed syscalls (and later garbage collection) or only because of the memory footprint? I don't expect the latter to be of any concern. From a performance point of view it doesn't make any difference.
Anyway, I went ahead and tested different scenarios with triclinic and orthorhombic boxes for the ADK protein system and compared in- and out-of-loop wrapping. The current approach seems fastest, but within error bars they're all the same speed. So it seems reasonable to go back to the single-coordinate in-loop wrapping.

The bottleneck is in fact the minimum image distance computation:
![image](https://user-images.githubusercontent.com/6999863/111917808-338a1100-8a82-11eb-916d-f8a653bd8cdf.png)
Perf log for a triclinic box. The minimum image computation is responsible for basically 100% of the `FastNS.search()` runtime.

![image](https://user-images.githubusercontent.com/6999863/111918570-fde72700-8a85-11eb-9854-74bac8860ad7.png)
Same for orthorhombic boxes, the minimum image computation causes almost 100% of the `FastNS.search()` runtime.

We should consider a lazy minimum image implementation leveraging the fact that we're always dealing with wrapped coordinates.

Reply:
I just tried out a lazy orthorhombic minimum image distance computation and it yields a speedup of about 22 %.
In that scenario the in-loop wrapping is a tiny bit faster so that's the way to go.
So, thank you for insisting! :smile: :+1: